dynamiqs/qarray.py

```diff
@@ -0,0 +1,554 @@
+        """
+
+    @abstractmethod
+    def ptrace(self, keep_dims: tuple[int, ...]) -> QArray:
+        """Partial trace of the quantum state.
+
+        Args:
+            keep_dims: Dimensions to keep.
+
+        Returns:
+            The partial trace of the quantum state.
+        """
+
+    @abstractmethod
+    def dag(self) -> QArray:
+        """Dagger of the quantum state.
+
+        Returns:
+            The dagger of the quantum state.
+        """
+
+    @abstractmethod
+    def isket(self) -> bool:
+        """Check if the quantum state is a ket.
+
+        Returns:
+            True if the quantum state is a ket, False otherwise.
+        """
+
+    @abstractmethod
+    def isbra(self) -> bool:
+        """Check if the quantum state is a bra.
+
+        Returns:
+            True if the quantum state is a bra, False otherwise.
+        """
+
+    @abstractmethod
+    def isdm(self) -> bool:
+        """Check if the quantum state is a density matrix.
+
+        Returns:
+            True if the quantum state is a density matrix, False otherwise.
+        """
+
+    def isop(self) -> bool:
+        """Check if the quantum state is an operator.
+
+        Returns:
+            True if the quantum state is an operator, False otherwise.
+        """
+        return self.isdm()
+
+    @abstractmethod
+    def isherm(self) -> bool:
+        """Check if the quantum state is Hermitian.
+
+        Returns:
+            True if the quantum state is Hermitian, False otherwise.
+        """
+
+    @abstractmethod
+    def toket(self) -> QArray:
+        """Convert the quantum state to a ket.
+
+        Returns:
+            The ket representation of the quantum state.
+        """
+
+    @abstractmethod
+    def tobra(self) -> QArray:
+        """Convert the quantum state to a bra.
+
+        Returns:
+            The bra representation of the quantum state.
+        """
+
+    @abstractmethod
+    def todm(self) -> QArray:
+        """Convert the quantum state to a density matrix.
+
+        Returns:
+            The density matrix representation of the quantum state.
+        """
+
+    def toop(self) -> QArray:
+        """Convert the quantum state to an operator.
+
+        Returns:
+            The operator representation of the quantum state.
+        """
+        return self.todm()
+
+    def proj(self) -> QArray:
+        """Projector of the quantum state.
+
+        Returns:
+            The projector of the quantum state.
+        """
+
+    # conversion methods
+    @abstractmethod
+    def to_numpy(self) -> np.ndarray:
+        """Convert the quantum state to a NumPy array.
+
+        Returns:
+            The NumPy array representation of the quantum state.
+        """
+
+    def to_qutip(self) -> qt.QObj:
+        """Convert the quantum state to a QuTiP object.
+
+        Returns:
+            A QuTiP object representation of the quantum state.
+        """
+        return qt.Qobj(self.to_numpy(), dims=self.dims)
```

Comment:
This should be implemented in the children classes, because the dense version will return a dense QObj, while the sparse version will return a sparse QObj.